# **Question 1: Loading the dataset (Churn_Modelling.csv) into a Pandas DataFrame.**

## **1.1 Implenting line wrapping for results**

In [ ]:
from IPython.display import HTML, display

def set_css():
  display(HTML('''
  <style>
    pre{
      white-space: pre-wrapped;
    }
  </style>
  '''))
get_ipython().events.register('pre_run_cell', set_css)

## **1.2 Importing dataset from device memory and loading into a Pandas DataFrame** 

In [ ]:
# Import dataset
import pandas as pd
DATA_DIR_1 = "/content/Churn_Modelling.csv"
churn_df = pd.read_csv (DATA_DIR_1, header=0)

The directory path of the dataset is stored in variable "DATA_DIR_1" and then stored in "churn_df" as a dataframe using the Pandas function "read_csv()"

# **Question 2: Show what kind of libraries need to be imported to pre-process this dataset. Explain the function for those libraries.**

In [ ]:
# Importing required libraries 
import numpy as np
import matplotlib as mpl

In order to pre-process the churn_modelling dataset, 3 librairies are needed for the follwing reasons respectively.
- **Pandas**: To store the dataset in a data structure that allows us to manipulate and analyse said dataset.
- **NumPy**: To perform scientific calculations on our data.
- **Matplotlib**: To visually represent the data as 2D charts.

# **Question 3**

Using Data Pre-processing and Data Cleansing technique, find the following information

## **3a) Number of customers (use python function)**

In [ ]:
# Find the number of customers
len(churn_df)

10000

Using the Pandas' len() function, we can find that there are 10000 rows in the dataset, hence 10000 customers.

## **3b) Number of variables/attributes**

In [ ]:
# Find the number of attributes 
len(churn_df.columns)

14

Using the Pandas' len() function with churn_df.columns as parameter, we can find that there are 14 columns in the dataset, hence 14 attributes/variables.

## **3c) Number of rows and columns**

In [ ]:
# Finding the number of rows and columns in the dataset
churn_df.shape 

(10000, 14)

Using the shape attribute of the dataframe, we can determine that there are **10000 rows** and **14 columns** from the output as the **rows** are given on the **left** and **columns** on the **right**.

## **3d) Determine the datatype of each column**

In [ ]:
# Finding the datatype of each column
print(churn_df.dtypes)

RowNumber            int64
CustomerId           int64
Surname             object
CreditScore          int64
Geography           object
Gender              object
Age                  int64
Tenure               int64
Balance            float64
NumOfProducts        int64
HasCrCard            int64
IsActiveMember       int64
EstimatedSalary    float64
Exited               int64
dtype: object


Using the dtypes attribue of the dataframe, we can output the data type of each column as shown above.

## **3e) Display basic statistic of all columns**

In [ ]:
churn_df.describe()

RowNumber    CustomerId   CreditScore           Age        Tenure  \
count  10000.00000  1.000000e+04  10000.000000  10000.000000  10000.000000   
mean    5000.50000  1.569094e+07    650.528800     38.921800      5.012800   
std     2886.89568  7.193619e+04     96.653299     10.487806      2.892174   
min        1.00000  1.556570e+07    350.000000     18.000000      0.000000   
25%     2500.75000  1.562853e+07    584.000000     32.000000      3.000000   
50%     5000.50000  1.569074e+07    652.000000     37.000000      5.000000   
75%     7500.25000  1.575323e+07    718.000000     44.000000      7.000000   
max    10000.00000  1.581569e+07    850.000000     92.000000     10.000000   

             Balance  NumOfProducts    HasCrCard  IsActiveMember  \
count    6383.000000   10000.000000  10000.00000    10000.000000   
mean    75345.698833       1.530200      0.70550        0.515100   
std     59818.076920       0.581654      0.45584        0.499797   
min        12.170000       1.000000      0.00000        0.000000   
25%     12964.305000       1.000000      0.00000        0.000000   
50%     85225.210000       1.000000      1.00000        1.000000   
75%    128357.590000       2.000000      1.00000        1.000000   
max    238387.560000       4.000000      1.00000        1.000000   

       EstimatedSalary        Exited  
count     10000.000000  10000.000000  
mean     100090.239881      0.203700  
std       57510.492818      0.402769  
min          11.580000      0.000000  
25%       51002.110000      0.000000  
50%      100193.915000      0.000000  
75%      149388.247500      0.000000  
max      199992.480000      1.000000

As shown above, the describe() function is used to output the count, mean, standard deviation, minimum value, quartiles and maximum value of each column.

## **3f) Balance attribute has many missing values. How do you propose to solve this problem? Show your steps.**

**Step 1:** Find the number of missing values in Balance column

In [ ]:
# Output the number of missing values in Balance column
print("Number of missing values in Balance column: %d" %churn_df['Balance'].isna().sum())

Number of missing values in Balance column: 3617


Using the isna() and sum() functions, we are able to determine that there is a total number of 3617 missing values in the Balance column.

**Step 2:** Calculate the mean of Balance Column

In [ ]:
# Calculate mean of balance column
mean_balance = churn_df.Balance.mean()
print("Mean Balance: %.2f" %mean_balance)

Mean Balance: 75345.70


Due to the high number of missing values, using the mean() function, we can calculate the mean balance, i.e. 75345.70 and use this value to replace the missing values.

**Step 3:** Replace the missing values with the mean balance value using the fillna() function.

In [ ]:
# Replace missing values with mean value
churn_df.Balance.fillna(mean_balance, inplace = True)
print("Number of missing values in Balance column: %d" %churn_df['Balance'].isna().sum())

Number of missing values in Balance column: 0


As shown in the output above, the number of missing values in the Balance column is zero after imputing it with its mean value.

## **3g) Can you use Mode to solve this problem with Balance attribute? Why?**

In [ ]:
# Calculating mode of Balance column
mode_balance = churn_df.Balance.mode()[0]
print("Balance mode: %.2f" %mode_balance)

Balance mode: 1317.82


Yes, the mode can be used in the case of the Balance column as mode, mean and median values can be used for data imputation for columns with numerical data, as long the the imputed value is not an outlier.

## **3h)Calculate how many customers has Credit Score below 650?**

In [ ]:
# Calculate number of customers with credit score less that 650
print("Number of customers with a credit score of less than 650: %d" %len(churn_df[churn_df.CreditScore < 650]))

Number of customers with a credit score of less than 650: 4900


As shown in the output above, by using the len() method, we can determine that there are 4900 customers with a credit score lower than 650.

## **3i) Calculate how many have exited the bank?**

In [ ]:
# Calculate how many customers have exited
print("Number who have exited: %d" %len(churn_df[(churn_df.Exited == 1) & (churn_df.IsActiveMember == 0)]))

Number who have exited: 1302


According to the meta data provided, it can be assumed that a customer is considered to have left only when both inactive (0) and when they have closed account with the bank (1). 
On that assumption, the number of customers who have exited calculated using the len() function is 1302.